File: src/main/scala/net/michalsitko/kloc/game/Rook.scala
package net.michalsitko.kloc.game

/**
 * Created with IntelliJ IDEA.
 * User: michal
 * Date: 5/25/13
 * Time: 8:15 PM
 * To change this template use File | Settings | File Templates.
 */

abstract trait Rook extends Piece {
  def isRookMove(chessboard: Chessboard, from: Field, to: Field): Boolean = {
    (from.sameRow(to) || from.sameColumn(to)) && !chessboard.somethingBetween(from, to)
  }

  def checkMoveCorrect(chessboard: Chessboard, move: Move, gameState: GameState): Boolean = {
    isRookMove(chessboard, move.from, move.to)
  }

  def getDirections(): List[(Int, Int)] = Rook.getDirections()
}

object Rook{
  def getDirections(): List[(Int, Int)] = {
    List((0, 1), (0, -1), (1, 0), (-1, 0))
  }
}

object RookFactory extends PieceFactory {
  def forColor(color: Color): Rook = {
    color match {
      case White() => WhiteRook
      case Black() => BlackRook
    }
  }
}

case object WhiteRook extends Rook {
  def getSymbol(): Char = 'R'

  def getColor(): Color = new White
}

case object BlackRook extends Rook {
  def getSymbol(): Char = WhiteRook.getSymbol().toLower

  def getColor(): Color = new Black
}

File: web/public/javascripts/game-rules/rook.js
define(['piece', 'color', 'chessboardUtils', 'move'], function(Piece, Color, ChessboardUtils, Move){
    var Rook = function(color) {
        Piece.call(this, color);
        this.symbol = 'r';
    };

    Rook.prototype = Object.create(Piece.prototype);

    Rook.prototype.isLegalMove = function(chessboard, move, gameState) {
        return (move.from.sameRow(move.to) || move.from.sameColumn(move.to)) && !chessboard.somethingBetween(move.from, move.to);
    }

    Rook.prototype.applyMove = function(chessboard, move, gameState) {
        var firstRow = this.color === Color.white ? 0 : 7;
        if(move.from.row === firstRow && move.from.column === 0){
            gameState.forColor(this.color).longCastlingEnabled = false;
        }

        if(move.from.row === firstRow && move.from.column === 7){
            gameState.forColor(this.color).shortCastlingEnabled = false;
        }

        return gameState;
    }

    Rook.prototype.vectors = function() {
        var Vector = ChessboardUtils.Vector;
        return [new Vector(0, 1), new Vector(0, -1), new Vector(1, 0), new Vector(-1, 0)];
    }

    return Rook;
});
